storage/src/journal/contiguous/variable.rs

```diff
@@ -0,0 +1,1441 @@
+    }
+
+    /// Read the item at the given position.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [Error::ItemPruned] if the item at `position` has been pruned.
+    /// - Returns [Error::ItemOutOfRange] if `position` is beyond the journal size.
+    /// - Returns other errors if storage or decoding fails.
+    pub async fn read(&self, position: u64) -> Result<V, Error> {
+        // Check bounds
+        if position >= self.size {
+            return Err(Error::ItemOutOfRange(position));
+        }
+
+        if position < self.oldest_retained_pos {
+            return Err(Error::ItemPruned(position));
+        }
+
+        // Read location from index
+        let location = self.locations.read(position).await?;
+
+        // Read item from data journal using the location
+        self.data.get(location.section, location.offset).await
+    }
+
+    /// Sync all pending writes to storage.
+    ///
+    /// This syncs both the data journal and the locations journal.
+    pub async fn sync(&mut self) -> Result<(), Error> {
+        // Sync all sections in the data journal
+        for &section in self.data.blobs.keys() {
+            self.data.sync(section).await?;
+        }
+
+        // Sync locations journal
+        self.locations.sync().await?;
+
+        Ok(())
+    }
+
+    /// Close the journal, syncing all pending writes.
+    ///
+    /// This closes both the data journal and the locations journal.
+    pub async fn close(mut self) -> Result<(), Error> {
+        self.sync().await?;
+        self.locations.close().await?;
+        self.data.close().await
+    }
+
+    /// Remove any underlying blobs created by the journal.
+    ///
+    /// This destroys both the data journal and the locations journal.
+    pub async fn destroy(self) -> Result<(), Error> {
+        self.locations.destroy().await?;
+        self.data.destroy().await
+    }
+
+    /// Return the section number where the next append will write.
+    const fn current_section(&self) -> u64 {
+        position_to_section(self.size, self.oldest_retained_pos, self.items_per_section)
+    }
+
+    /// Repair the locations journal to match the data journal.
+    ///
+    /// The data journal is the source of truth. This function scans it to determine
+    /// what SHOULD be in the locations journal, then fixes any mismatches.
+    ///
+    /// # Returns
+    ///
+    /// Returns `(size, oldest_retained_pos)`.
+    async fn validate_and_repair_locations(
+        data: &variable::Journal<E, V>,
+        locations: &mut fixed::Journal<E, Location>,
+        items_per_section: u64,
+    ) -> Result<(u64, u64), Error> {
+        // === Step 1: Scan data journal to determine expected state ===
+
+        let (expected_locations_size, expected_locations_oldest_pos) = if data.blobs.is_empty() {
+            // No data blobs → journal is empty or fully pruned
+            // Size comes from locations journal.
+            (locations.size().await?, locations.size().await?)
```

Comment:
Note that it's not possible to distinguish these scenarios:

1. Data partition was errantly deleted (corruption)
2. Journal is empty
3. Journal crashed while pruning to empty -- data journal was emptied but locations wasn't

The data journal is authoritative, but the _lack_ of a data journal can mean several things happened.

We handle all of these cases by using whatever is in the locations journal.